File: go/leetcode/linkedlist_test.go
package leetcode

import (
	"errors"
	"testing"

	"github.com/stretchr/testify/assert"
)

var ErrIndexOutOfRange = errors.New("index out of range")

type node struct {
	val  int
	prev *node
	next *node
}

type MyLinkedList struct {
	count int
	head  *node
	tail  *node
}

func Constructor() MyLinkedList {
	return MyLinkedList{}
}

func (ths *MyLinkedList) Get(index int) (int, error) {
	node := ths.getNode(index)
	if node == nil {
		return -1, ErrIndexOutOfRange
	} else {
		return node.val, nil
	}
}

func (ths *MyLinkedList) AddAtHead(val int) {
	head := ths.head
	node := &node{val: val, next: head}
	ths.head = node
	if ths.tail == nil {
		ths.tail = node
	}
	if head != nil {
		head.prev = node
	}
	ths.count += 1
}

func (ths *MyLinkedList) AddAtTail(val int) {
	tail := ths.tail
	if tail == nil {
		ths.AddAtHead(val)
	} else {
		node := &node{val: val, prev: tail}
		ths.tail = node
		tail.next = node
		ths.count += 1
	}
}

// [0, ...]
func (ths *MyLinkedList) getNode(index int) *node {
	if index < 0 {
		return nil
	}
	node := ths.head
	for i := 0; i < index; i++ {
		if nil == node {
			break
		}
		node = node.next
	}
	return node
}

// 在链表中的第 index 个节点之前添加值为 val  的节点。如果 index 等于链表的长度，则该节点将附加到链表的末尾
func (ths *MyLinkedList) AddAtIndex(index int, val int) {
	if index <= 0 {
		ths.AddAtHead(val)
		return
	} else if index == ths.count {
		ths.AddAtTail(val)
		return
	} else if index > ths.count {
		return
	}
	before := ths.getNode(index)
	node := &node{val: val, prev: before.prev, next: before}
	before.prev.next = node
	before.prev = node
	ths.count += 1
}

func (ths *MyLinkedList) DeleteAtIndex(index int) {
	node := ths.getNode(index)
	if node == nil {
		return
	}
	if ths.tail == node {
		ths.tail = node.prev
	}
	if ths.head == node {
		ths.head = node.next
	}
	if node.next != nil {
		node.next.prev = node.prev
	}
	if node.prev != nil {
		node.prev.next = node.next
	}
	ths.count -= 1
}

func TestMyLinkedList(t *testing.T) {
	l := Constructor() //[]
	v, e := l.Get(0)
	if assert.Error(t, e) {
		assert.Equal(t, ErrIndexOutOfRange, e)
	}
	assert.Equal(t, -1, v)
	l.AddAtHead(1)     //[1]
	l.AddAtTail(3)     //[1,3]
	l.AddAtIndex(1, 2) //[1,2,3]
	v, e = l.Get(-1)
	if assert.Error(t, e) {
		assert.Equal(t, ErrIndexOutOfRange, e)
	}
	assert.Equal(t, -1, v)
	v, e = l.Get(0)
	assert.Nil(t, e)
	assert.Equal(t, 1, v)
	v, e = l.Get(1)
	assert.Nil(t, e)
	assert.Equal(t, 2, v)
	v, e = l.Get(2)
	assert.Nil(t, e)
	assert.Equal(t, 3, v)
	v, e = l.Get(3)
	if assert.Error(t, e) {
		assert.Equal(t, ErrIndexOutOfRange, e)
	}
	assert.Equal(t, -1, v)
	l.DeleteAtIndex(1) //[1,3]
	v, e = l.Get(1)
	assert.Nil(t, e)
	assert.Equal(t, 3, v)
	v, e = l.Get(2)
	if assert.Error(t, e) {
		assert.Equal(t, ErrIndexOutOfRange, e)
	}
	assert.Equal(t, -1, v)
	l.DeleteAtIndex(0) //[3]
	v, e = l.Get(0)
	assert.Nil(t, e)
	assert.Equal(t, 3, v)
	l.DeleteAtIndex(0) //[]
	v, e = l.Get(0)
	if assert.Error(t, e) {
		assert.Equal(t, ErrIndexOutOfRange, e)
	}
	assert.Equal(t, -1, v)
	l.AddAtTail(3) //[3]
	v, e = l.Get(0)
	assert.Nil(t, e)
	assert.Equal(t, 3, v)
	l.AddAtIndex(1, 4) //[3,4]
	v, e = l.Get(0)
	assert.Nil(t, e)
	assert.Equal(t, 3, v)
	v, e = l.Get(1)
	assert.Nil(t, e)
	assert.Equal(t, 4, v)
}

func TestMyLinkedList2(t *testing.T) {
	l := Constructor() //[]
	l.AddAtIndex(0, 10)
	l.AddAtIndex(0, 20)
	l.AddAtIndex(0, 30)
	v, e := l.Get(0)
	assert.Nil(t, e)
	assert.Equal(t, 30, v)
	v, e = l.Get(1)
	assert.Nil(t, e)
	assert.Equal(t, 20, v)
	v, e = l.Get(2)
	assert.Nil(t, e)
	assert.Equal(t, 10, v)
}

File: rust/leetcode/tests/linkedlist_test.rs
use std::{
    cell::RefCell,
    rc::{Rc, Weak},
};

type RefNode = RefCell<Node>;

/**
 * Your MyLinkedList object will be instantiated and called as such:
 * let obj = MyLinkedList::new();
 * let ret_1: i32 = obj.get(index);
 * obj.add_at_head(val);
 * obj.add_at_tail(val);
 * obj.add_at_index(index, val);
 * obj.delete_at_index(index);
 */
#[derive(Default)]
pub struct MyLinkedList {
    count: i32,
    head: Option<Rc<RefNode>>,
    tail: Option<Rc<RefNode>>,
}

struct Node {
    val: i32,
    prev: Option<Weak<RefNode>>,
    next: Option<Rc<RefNode>>,
}

impl Node {
    fn new(val: i32, prev: Option<Weak<RefNode>>, next: Option<Rc<RefNode>>) -> Self {
        Self { val, prev, next }
    }
}

/**
 * `&self` means the method takes an immutable reference.
 * If you need a mutable reference, change it to `&mut self` instead.
 */
impl MyLinkedList {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn get(&self, index: i32) -> i32 {
        if let Some(n) = self.get_node(index) {
            n.as_ref().borrow().val
        } else {
            -1
        }
    }

    pub fn add_at_head(&mut self, val: i32) {
        let n = Rc::new(RefCell::new(Node::new(val, None, self.head.clone())));
        if self.tail.is_none() {
            self.tail = Some(n.clone());
        }
        if let Some(head) = &self.head {
            head.as_ref().borrow_mut().prev = Some(Rc::downgrade(&n));
        }
        self.head = Some(n);
        self.count += 1;
    }

    pub fn add_at_tail(&mut self, val: i32) {
        if let Some(tail) = &self.tail {
            let n = Rc::new(RefCell::new(Node::new(
                val,
                Some(Rc::downgrade(tail)),
                None,
            )));
            tail.as_ref().borrow_mut().next = Some(n.clone());
            self.tail = Some(n);
            self.count += 1;
        } else {
            self.add_at_head(val);
        }
    }

    pub fn add_at_index(&mut self, index: i32, val: i32) {
        if index <= 0 {
            self.add_at_head(val);
        } else if index == self.count {
            self.add_at_tail(val);
        } else if index > self.count {
        } else {
            let before = self.get_node(index).unwrap();
            let prev = before.as_ref().borrow().prev.clone();
            let n = Rc::new(RefCell::new(Node::new(
                val,
                prev.clone(),
                Some(before.clone()),
            )));
            before.as_ref().borrow_mut().prev = Some(Rc::downgrade(&n));
            prev.as_ref()
                .unwrap()
                .upgrade()
                .unwrap()
                .as_ref()
                .borrow_mut()
                .next = Some(n);
            self.count += 1;
        }
    }

    pub fn delete_at_index(&mut self, index: i32) {
        if let Some(n) = self.get_node(index) {
            if let Some(head) = self.head.as_ref() {
                if Rc::ptr_eq(head, &n) {
                    self.head = n.as_ref().borrow().next.clone();
                }
            }
            if let Some(tail) = self.tail.as_ref() {
                if Rc::ptr_eq(tail, &n) {
                    self.tail = n
                        .as_ref()
                        .borrow()
                        .prev
                        .as_ref()
                        .map(|x| x.upgrade().unwrap());
                }
            }
            if let Some(next) = &n.as_ref().borrow().next {
                next.as_ref().borrow_mut().prev = n.as_ref().borrow().prev.clone();
            }
            if let Some(prev) = n
                .as_ref()
                .borrow()
                .prev
                .as_ref()
                .map(|x| x.upgrade().unwrap())
            {
                prev.as_ref().borrow_mut().next = n.as_ref().borrow().next.clone();
            }
            self.count -= 1
        }
    }

    fn get_node(&self, index: i32) -> Option<Rc<RefNode>> {
        if index >= self.count || index < 0 {
            return None;
        }
        let mut node = self.head.clone();
        for _ in 0..index {
            if let Some(n) = node {
                node = n.as_ref().borrow().next.clone();
            } else {
                break;
            }
        }
        node
    }
}

#[cfg(test)]
mod tests {
    use crate::MyLinkedList;

    #[test]
    fn it_works1() {
        let mut l = MyLinkedList::new(); //[]
        let v = l.get(0);
        assert_eq!(-1, v);
        l.add_at_head(1); //[1]
        l.add_at_tail(3); //[1,3]
        l.add_at_index(1, 2); //[1,2,3]
        let v = l.get(-1);
        assert_eq!(-1, v);
        let v = l.get(0);
        assert_eq!(1, v);
        let v = l.get(1);
        assert_eq!(2, v);
        let v = l.get(2);
        assert_eq!(3, v);
        let v = l.get(3);
        assert_eq!(-1, v);
        l.delete_at_index(1); //[1,3]
        let v = l.get(1);
        assert_eq!(3, v);
        let v = l.get(2);
        assert_eq!(-1, v);
        l.delete_at_index(0); //[3]
        let v = l.get(0);
        assert_eq!(3, v);
        l.delete_at_index(0); //[]
        let v = l.get(0);
        assert_eq!(-1, v);
        l.add_at_tail(3); //[3]
        let v = l.get(0);
        assert_eq!(3, v);
        l.add_at_index(1, 4); //[3,4]
        let v = l.get(0);
        assert_eq!(3, v);
        let v = l.get(1);
        assert_eq!(4, v);
    }

    #[test]
    fn it_works2() {
        let mut l = MyLinkedList::new(); //[]
        l.add_at_index(0, 10);
        l.add_at_index(0, 20);
        l.add_at_index(0, 30);
        let v = l.get(0);
        assert_eq!(30, v);
        let v = l.get(1);
        assert_eq!(20, v);
        let v = l.get(2);
        assert_eq!(10, v);
    }
}
